File: R/forLater/col_line.R
#' Plot coloured line(s) in 2 dimensions in the current figure.
#' 
#' @param x A vector or matrix of points on the horizontal axis.
#' @param y A vector or matrix of points on the vertical axis.
#' @param c A vector or matrix of values representing the colour to draw at each point.
#' @note x, y and c must all be the same size. If x, y, and c are matrices, one line is drawn for each column. The color axis will by default span the range of values in c, i.e., caxis will be c(min(min(c)), max(max(c))). This can be changed by calling caxis after colline.

col_line <- function(x, y, c) {
  if (length(x) == length(y) & length(x) == length(c)) {
     x <- matrix(x, nrow = 1)
     y <- matrix(y, nrow = 1)
     c <- matrix(c, nrow = 1)
     X <- matrix(0, length(x), 1)
     Y <- matrix(0, length(x), 1)
     C <- matrix(0, length(x), 1)
     for (k in 1:length(x)) {
       X[k] <- x[k]
       Y[k] <- y[k]
       C[k] <- c[k]
     }
     data <- list()
     for (j in 1:(length(X) - 1)) {
       for (l in (j + 1)) {
         data[[j]] <- data.frame(X[j:l], Y[j:l])
       }
     }
     plot(NA, xlim = c(0, max(x)), ylim = c(0, max(y)))
     for (i in 1:(length(X)-1)) {
       d <- data[[i]]
       lines(x = d[, 1], y = d[, 2], col = C[i], lwd = 3, xlab = NULL, ylab = NULL)
     }
  } else {
    if (nrow(x) == nrow(y) & nrow(x) == nrow(c) & ncol(x) == ncol(y) & ncol(x) == ncol(c)) {
      if (ncol(x) == 1) {
        x <- matrix(x, nrow = 1)
        y <- matrix(y, nrow = 1)
        c <- matrix(c, nrow = 1)
      }
      datax <- list()
      for (j in 1:(nrow(x) - 1)) {
        for (l in (j + 1)) {
          datax[[j]] <- data.frame(x[j:l, ])
        }
      }
      datay <- list()
      for (j in 1:(nrow(y) - 1)) {
        for (l in (j + 1)) {
          datay[[j]] <- data.frame(y[j:l, ])
        }
      }
      graphics::matplot(NA, xlim = c(0, max(x)), ylim = c(0, max(y)))
      for (i in 1:(nrow(x)-1)) {
        dx <- datax[[i]]
        dy <- datay[[i]]
        graphics::matlines(x = dx, y = dy, col = c[i], lwd = 3, lty = 1, xlab = NULL, ylab = NULL)
      }
    }
  }
}

File: matlab/tagtools/plotting/col_line.m
function    h = col_line(x,y,c)

%     h = col_line(x,y,c)
%     Plot coloured line(s) in 2 dimensions in the current figure.
%
%		Inputs:
%		x is a vector or matrix of points on the horizontal axis.
%		y is a vector or matrix of points on the vertical axis.
%		c is a vector or matrix of values representing the colour to draw
%		  at each point.

%		Result:
%		h is a vector of patch handles which can be used with 'set' to change
%		properties of the line. For example: set(h,'LineWidth',8) will make the
%		line wider.
%
%     x, y and c must all be the same size. If x, y, and c are matrices, 
%		one line is drawn for each column. The color axis will by default span the
%		range of values in c, i.e., caxis will be [min(min(c)) max(max(c))]. This
%		can be changed by calling caxis after colline.
%
%		Example:
%		 T = cumsum(randn(1000,2)) ;
%		 col_line(T(:,1),T(:,2),(1:size(T,1))') ;
% 	    Draws a a random walk in the current figure coloured by sample number.
%
%     Valid: Matlab, Octave
%     markjohnson@st-andrews.ac.uk
%     Last modified: 8 June 2017

if nargin<3,
	help col_line
	return
end
	
M = [size(x) size(y) size(c)] ;
if any(M==1),
   x = x(:) ;
   y = y(:) ;
   c = c(:) ;
end

h = zeros(size(x,2),1) ;
for k=1:size(x,2),
   h(k) = patch('xdata',[x(:,k);NaN],'ydata',[y(:,k);NaN],'cdata',[c(:,k);NaN],...
          'linestyle','-','edgecolor','flat');
end
set(h,'LineWidth',3) ;
